modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/statistic/SqlStatisticManagerImpl.java

```diff
@@ -171,6 +135,66 @@ public void start() {
                 tableSizeMap.putIfAbsent(table.id(), DEFAULT_VALUE);
             }
         }
+
+        scheduler.scheduleAtFixedRate(this::update, INITIAL_DELAY, REFRESH_PERIOD, TimeUnit.MILLISECONDS);
+    }
+
+    private void update() {
+        if (!latestUpdateFut.get().isDone()) {
+            return;
+        }
+
+        for (Map.Entry<Integer, ActualSize> ent : tableSizeMap.entrySet()) {
+            Integer tableId = ent.getKey();
+
+            if (droppedTables.contains(tableId)) {
+                continue;
+            }
+
+            TableViewInternal tableView = tableManager.cachedTable(tableId);
+
+            if (tableView == null) {
+                LOG.debug("No table found to update statistics [id={}].", ent.getKey());
+                continue;
+            }
+
+            CompletableFuture<Void> updateResult = statSupplier.estimatedSizeWithLastUpdate(tableView.internalTable())
+                    .handle((res, err) -> {
+                            if (err != null) {
+                                LOG.debug("Failed to update table statistics for [tableId={}].", err, tableId);
+
+                                return null;
+                            } else {
+                                ActualSize estimatedTableSize = new ActualSize(Math.max(res.keyLong(), DEFAULT_TABLE_SIZE), res.value());
+                                ActualSize prevSize = tableSizeMap.get(tableId);
+                                // the table can be concurrently dropped and we shouldn't put new value in this case.
+                                tableSizeMap.compute(tableId, (k, v) -> {
+                                    if (v == null) {
+                                        return estimatedTableSize;
+                                    }
+
+                                    // check for stale update
+                                    if (v.timestamp.compareTo(res.value()) > 0) {
+                                        return v;
+                                    }
+
+                                    return estimatedTableSize;
+                                });
+
+                                if (!estimatedTableSize.equals(prevSize)) {
+                                    StatisticUpdatesSupplier supplier = changesSupplier.get();
+                                    if (supplier != null) {
+                                        supplier.accept(tableId);
+                                    }
+                                }
```

Comment:
This is barely readable.
1. Why do we compare timestamps, even if size is the same? Need some explanation there.
2. What is happening in outer if-block? Do ask StatisticsManager to refresh statistics? even if we found a single entry was touched?